### ISSUE TITLE
`asPreview: true` breaks query

### ISSUE THREAD

**Issue by joernroeder** [wp-graphql/wp-graphql#2063]
I'm debugging a preview issue with my gatsby instance and noticed that they come back with an "internal server error" from wp-graphQL 1.6.2 on wp 5.7.2

I was able to trim down the generated query to the following and still see the error:

```graphql
query PREVIEW_QUERY($id: ID!) {
  post(id: $id, idType: ID, asPreview: true) {
    id
  }    
}
```

flipping `asPreview` to `false` works as expected and getting the result

```json
{
  "data": {
    "post": {
      "id": "cG9zdDoyMDUz"
    }
  },
  "extensions": {
    "debug": []
  }
}
```

```json
{
      "debugMessage": "Call to a member function then() on null",
      "message": "Internal server error",
      "extensions": {
        "category": "internal"
      },
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "post"
      ],
      "trace": [
        {
          "file": "/wp-content/plugins/wp-graphql/src/Utils/InstrumentSchema.php",
          "line": 151,
          "call": "WPGraphQL\\Type\\ObjectType\\RootQuery::WPGraphQL\\Type\\ObjectType\\{closure}(null, array(3), instance of WPGraphQL\\AppContext, instance of GraphQL\\Type\\Definition\\ResolveInfo)"
        },
        {
          "file": "/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/ReferenceExecutor.php",
          "line": 623,
          "call": "WPGraphQL\\Utils\\InstrumentSchema::WPGraphQL\\Utils\\{closure}(null, array(3), instance of WPGraphQL\\AppContext, instance of GraphQL\\Type\\Definition\\ResolveInfo)"
        },
        {
          "file": "/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/ReferenceExecutor.php",
          "line": 550,
          "call": "GraphQL\\Executor\\ReferenceExecutor::resolveFieldValueOrError(instance of GraphQL\\Type\\Definition\\FieldDefinition, instance of GraphQL\\Language\\AST\\FieldNode, instance of Closure, null, instance of GraphQL\\Type\\Definition\\ResolveInfo)"
        },
        {
          "file": "/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/ReferenceExecutor.php",
          "line": 1195,
          "call": "GraphQL\\Executor\\ReferenceExecutor::resolveField(GraphQLType: RootQuery, null, instance of ArrayObject(1), array(1))"
        },
        {
          "file": "/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/ReferenceExecutor.php",
          "line": 264,
          "call": "GraphQL\\Executor\\ReferenceExecutor::executeFields(GraphQLType: RootQuery, null, array(0), instance of ArrayObject(1))"
        },
        {
          "file": "/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/ReferenceExecutor.php",
          "line": 215,
          "call": "GraphQL\\Executor\\ReferenceExecutor::executeOperation(instance of GraphQL\\Language\\AST\\OperationDefinitionNode, null)"
        },
        {
          "file": "/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/Executor.php",
          "line": 156,
          "call": "GraphQL\\Executor\\ReferenceExecutor::doExecute()"
        },
        {
          "file": "/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/GraphQL.php",
          "line": 162,
          "call": "GraphQL\\Executor\\Executor::promiseToExecute(instance of GraphQL\\Executor\\Promise\\Adapter\\SyncPromiseAdapter, instance of WPGraphQL\\WPSchema, instance of GraphQL\\Language\\AST\\DocumentNode, null, instance of WPGraphQL\\AppContext, array(1), 'PREVIEW_QUERY', null)"
        },
        {
          "file": "/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Server/Helper.php",
          "line": 311,
          "call": "GraphQL\\GraphQL::promiseToExecute(instance of GraphQL\\Executor\\Promise\\Adapter\\SyncPromiseAdapter, instance of WPGraphQL\\WPSchema, instance of GraphQL\\Language\\AST\\DocumentNode, null, instance of WPGraphQL\\AppContext, array(1), 'PREVIEW_QUERY', null, array(29))"
        },
        {
          "file": "/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Server/Helper.php",
          "line": 211,
          "call": "GraphQL\\Server\\Helper::promiseToExecuteOperation(instance of GraphQL\\Executor\\Promise\\Adapter\\SyncPromiseAdapter, instance of GraphQL\\Server\\ServerConfig, instance of GraphQL\\Server\\OperationParams)"
        },
        {
          "file": "/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Server/StandardServer.php",
          "line": 136,
          "call": "GraphQL\\Server\\Helper::executeOperation(instance of GraphQL\\Server\\ServerConfig, instance of GraphQL\\Server\\OperationParams)"
        },
        {
          "file": "/wp-content/plugins/wp-graphql/src/Request.php",
          "line": 666,
          "call": "GraphQL\\Server\\StandardServer::executeRequest(instance of GraphQL\\Server\\OperationParams)"
        },
        {
          "file": "/wp-content/plugins/wp-graphql/src/Router.php",
          "line": 468,
          "call": "WPGraphQL\\Request::execute_http()"
        },
        {
          "file": "/wp-content/plugins/wp-graphql/src/Router.php",
          "line": 270,
          "call": "WPGraphQL\\Router::process_http_request()"
        },
        {
          "file": "/wp-includes/class-wp-hook.php",
          "line": 292,
          "call": "WPGraphQL\\Router::resolve_http_request(instance of WP)"
        },
        {
          "file": "/wp-includes/class-wp-hook.php",
          "line": 316,
          "call": "WP_Hook::apply_filters(null, array(1))"
        },
        {
          "file": "/wp-includes/plugin.php",
          "line": 551,
          "call": "WP_Hook::do_action(array(1))"
        },
        {
          "file": "/wp-includes/class-wp.php",
          "line": 388,
          "function": "do_action_ref_array('parse_request', array(1))"
        },
        {
          "file": "/wp-includes/class-wp.php",
          "line": 750,
          "call": "WP::parse_request((empty string))"
        },
        {
          "file": "/wp-includes/functions.php",
          "line": 1291,
          "call": "WP::main((empty string))"
        },
        {
          "file": "/wp-blog-header.php",
          "line": 16,
          "function": "wp()"
        },
        {
          "file": "/index.php",
          "line": 17,
          "function": "require('/wp-blog-header.php')"
        }
      ]
    }
  ],
  "data": {
    "post": null
  },
  "extensions": {
    "debug": []
  }
}
```

**Comment by jasonbahl** [comment]
@joernroeder Thanks for the issue. I'll see if I can reproduce 👀

**Comment by jasonbahl** [comment]
@joernroeder can you share more information about the post so I can reproduce? Is the post a published post? Wha

**Comment by jasonbahl** [comment]
@joernroeder very interesting. I'm able to reproduce this if I do the query using the `idType: ID` argument, but _not_ if I use `idType: DATABASE_ID` 🤔 

**USING DATABASE_ID**
![Screen Shot 2021-08-09 at 8 42 13 AM](https://user-images.githubusercontent.com/1260765/128725328-9bef6390-af3a-4cb0-8a61-805ca3b6181b.png)

**USING ID**
![Screen Shot 2021-08-09 at 8 43 35 AM](https://user-images.githubusercontent.com/1260765/128725372-0008f4dd-8c88-40b3-a9a9-56275d024600.png)

**Comment by joernroeder** [comment]
@jasonbahl good/bad that you're able to reproduce it. I'm using gatsby which also sends the ID not databaseId along. It was a post, still a draft at this point

**Comment by jasonbahl** [comment]
@joernroeder I have a PR in-progress that might fix this. 

First, I want to get more information to reproduce even more precisely. 

You passed id `cG9zdDoyMDUz` which is `post:2053`

Can you share as much info as possible about this post's raw data in your database?

Info such as: 

- post_author
- post_type
- post_status
- post_parent

Some of this information would be helpful in debugging. 

Maybe even a screenshot of the Post as it exists in the wp-admin?

And maybe a screenshot for how WordPress loads the post when you preview in vanilla WordPress (without Gatsby) by hitting `yoursite.com/?p=2053&preview=true`

**Comment by jasonbahl** [comment]
@joernroeder Can you see if applying the changes from this PR (https://github.com/wp-graphql/wp-graphql/pull/2064) resolves the issue for you?

**Comment by joernroeder** [comment]
@jasonbahl I've applied the patch https://github.com/wp-graphql/wp-graphql/pull/2064/files#diff-d0159420813e3027d728124c9f2cdd153dcedec077e2de9f60669befba20e328R667 and now the query doesn't throw an error anymore 🥳.

But it also does not return the post but `null`, not sure if this is the expected behavior:

```json
{
  "data": {
    "post": null
  },
  "extensions": {
    "debug": []
  }
}
```

**Comment by jasonbahl** [comment]
@joernroeder Can you share more info about the post with ID 2053, as requested here (https://github.com/wp-graphql/wp-graphql/issues/2063#issuecomment-895305729)? 

This will help me debug better.

**Comment by jasonbahl** [comment]
@joernroeder in my testing, passing an ID of a post that's _already_ a revision was retuning the error, but passing an ID of a post that _has_ revisions was resolving properly. 

For example, if I create a new post (for me, post ID `3227`) then make a change and try to preview it, my change will be stored as a post of the `revision` post_type, with `3227` as the `post_parent` value: 

![Screen Shot 2021-08-10 at 2 26 51 PM](https://user-images.githubusercontent.com/1260765/128930414-0cfbf7e9-464e-4c1b-befa-af05315b616a.png)

I can preview my post by passing the ID of the post (`3227`) to the url, and in response WordPress will get content from the revision post (`3230`) and merge it with the preview template and I can see my changes. 

WPGraphQL does the same, you can pass ID `3227` and ask for a preview, and you will see the pending changes, resolved under the hood from post `3230`: 

![Screen Shot 2021-08-10 at 2 29 57 PM](https://user-images.githubusercontent.com/1260765/128930647-c6ad79a9-a362-44cd-91f5-40cd8196368f.png)

Now, in WPGraphQL, if I pass ID `3230` (the revision), this is when the error you reported was occurring (from what I can tell):

![Screen Shot 2021-08-10 at 2 32 47 PM](https://user-images.githubusercontent.com/1260765/128931096-31e98f9b-61ed-4474-b7e6-fef01606c237.png)


but now, with this PR, returns null: 

![Screen Shot 2021-08-10 at 2 33 13 PM](https://user-images.githubusercontent.com/1260765/128931141-45bcd27d-3235-44d8-89d4-757c172b5db9.png)


In WordPress core, if you attempt to preview a post using the revision ID, you will get a `wp_die()` message, like so: 

![Screen Shot 2021-08-10 at 2 23 40 PM](https://user-images.githubusercontent.com/1260765/128929978-7d33f12c-085f-4055-bb93-1d2faa872509.png)

So, I believe WPGraphQL returning a `null` is on par with (and friendlier than) WordPress core's behavior of not returning the post you asked for, given the ID that was passed to it. 

Now, the main question remains, the ID that is being passed for you, `2053`, is that of the post_type `revision`? 

This _might_ be a Gatsby Source WordPress bug where it's passing revision IDs for preview instead of passing the ID of the post you're trying to preview 🤔

**Comment by joernroeder** [comment]
@jasonbahl 2053 is the Id of the post not a revision. I'm also seeing it in the edit url `/wp-admin/post.php?post=2053&action=edit`

**Comment by joernroeder** [comment]
@jasonbahl I'll try to verify via the db, not sure if I have access.

**Comment by joernroeder** [comment]
Info from the wp_posts table:

- post_author: `24`
- post_type: `post`
- post_status: `inherit` (It's now published, but wasn't when I encounter the issue)
- post_parent: `0`

**Comment by jasonbahl** [comment]
@joernroeder the only way I've been able to reproduce the bug (which I fixed with #2064), is to pass an ID that can't be previewed. 

Are you able to reproduce consistently? 

If so, can you share step-by-step instructions on how you're creating posts that end up in a state where the preview errors. 

For me, here's what I tried: 

## Scenario 1: Click "Preview" on a draft post, use draft ID in GraphQL

- In WordPress, I created a new Post
- I saved the post as a draft
- I clicked "Preview" (this ensures WordPress creates a proper revision to preview)
- This took me to the URL `http://wpgraphql.local/?p=3232&preview=true`
- I used the ID of the draft (`3232`) in the following GraphQL Query: 

```graphql
query PreviewPost($id: ID!) {
  post(id: $id, idType: DATABASE_ID, asPreview: true) {
    id
    databaseId
    isPreview
    content
  }
}
```

### Authenticated
In an authenticated GraphiQL client (GraphiQL in the WP Admin) I see the following (expected) results: 

![Screen Shot 2021-08-13 at 8 12 04 AM](https://user-images.githubusercontent.com/1260765/129370249-926b33a4-d51d-4b92-8b91-dd4e64de9545.png)


### Public (non-authenticated)
In a non-authenticated GraphiQL, I see the following (also expected) results:

![Screen Shot 2021-08-13 at 8 23 02 AM](https://user-images.githubusercontent.com/1260765/129371788-0deb00bd-14c5-4009-916e-622458c0a13d.png)




## Scenario 2: Use Draft ID in GraphQL without clicking Preview

- In WordPress, I created a new Post
- I saved the post as a draft
- After the save, the ID is generated and shown in the URL: `http://wpgraphql.local/wp-admin/post.php?post=&action=edit&classic-editor`
- I did not click "preview" this time, but still took the ID and used it in the same query as above

### Authenticated
In an authenticated GraphiQL client (GraphiQL in the WP Admin) I see the following (expected) results: 

![Screen Shot 2021-08-13 at 8 15 10 AM](https://user-images.githubusercontent.com/1260765/129370660-c7537261-4cac-4c86-a6e7-2f35e827aadb.png)

### Public (non-authenticated)
In a non-authenticated GraphiQL, I see the following (also expected) results:

![Screen Shot 2021-08-13 at 8 21 52 AM](https://user-images.githubusercontent.com/1260765/129371660-963a713a-3261-4265-a8ae-6a7fd9d8d4b1.png)


----


In both of these scenarios, both as a public and authenticated user, I see results I would expect. 

Now, the only time I've been able to reproduce the bug you posted originally, is to pass an ID of a post that cannot be previewed. 

For example, we can see in the results of the last authenticated query, that if we preview post `3236`, the ID of the preview node is `3237`. 

If I take that ID and try to preview _it_ then I do get the error you posted:

#### WPGraphQL v1.6.2

![Screen Shot 2021-08-13 at 8 19 04 AM](https://user-images.githubusercontent.com/1260765/129371197-e9349140-7292-4f7b-a91c-642009176898.png)

#### PR #2064 

With the PR applied, even with an ID that can't be previewed, I now do not get any errors, just a null, which is what I believe should be expected from a Graph. 

![Screen Shot 2021-08-13 at 8 20 06 AM](https://user-images.githubusercontent.com/1260765/129371340-feb2e339-0f05-4781-b2a0-fb0f42e25419.png)

![Screen Shot 2021-08-13 at 8 20 30 AM](https://user-images.githubusercontent.com/1260765/129371512-b486de9f-2246-404b-8c23-244d8b01aa24.png)

**Comment by jasonbahl** [comment]
@joernroeder I've merged #2064 as I believe it resolves the original issue. 

If you have detailed steps on how to reproduce, including maybe screenshots or a video or something, that would be super helpful. If you do have those details, please open a new issue and mention this issue for context.

**Comment by joernroeder** [comment]
Thanks @jasonbahl! I'll give it shot and report back to you. Haven't had the time to play with it in full detail yet.
